utilities/transactions/pessimistic_transaction.cc

```diff
@@ -551,25 +557,29 @@ Status PessimisticTransaction::TryLock(ColumnFamilyHandle* column_family,
 // transaction snapshot_.
 Status PessimisticTransaction::ValidateSnapshot(
     ColumnFamilyHandle* column_family, const Slice& key,
-    SequenceNumber prev_seqno, SequenceNumber* new_seqno) {
+    SequenceNumber tracked_at_seq, SequenceNumber* new_tracked_at_seq) {
   assert(snapshot_);
 
-  SequenceNumber seq = snapshot_->GetSequenceNumber();
-  if (prev_seqno <= seq) {
+  SequenceNumber snap_seq = snapshot_->GetSequenceNumber();
+  if (tracked_at_seq <= snap_seq) {
     // If the key has been previous validated at a sequence number earlier
     // than the curent snapshot's sequence number, we already know it has not
     // been modified.
     return Status::OK();
   }
+  // Otherwise we have either
+  // 1: tracked_at_seq == kMaxSequenceNumber, i.e., first time tracking the key
+  // 2: snap_seq < tracked_at_seq: last time we lock the key was via
+  // skip_validate option which means we had skipped ValidateSnapshot In both
+  // cases we should do ValidateSnapshot now.
 
-  *new_seqno = seq;
+  *new_tracked_at_seq = snap_seq;
 
   ColumnFamilyHandle* cfh =
       column_family ? column_family : db_impl_->DefaultColumnFamily();
 
-  return TransactionUtil::CheckKeyForConflicts(db_impl_, cfh, key.ToString(),
-                                               snapshot_->GetSequenceNumber(),
-                                               false /* cache_only */);
+  return TransactionUtil::CheckKeyForConflicts(
```

Comment:
I'm wondering why don't we let `TransactionUtil::CheckKeyForConflicts` return the last written sequence number of the key, so that the next time we have `snap_seq < tracked_at_seq` we don't need to double check. (Does it make sense?)

Reply:
If it there is a conflict it unlocks the key and returns busy. the txn cannot continue with a conflict and should naturally rollback.

Reply:
I also added more comments to clarify what tracked_at_seq signifies.